library/lmots.c

```diff
@@ -181,12 +182,17 @@ static int create_digit_array_with_checksum( const mbedtls_lmots_parameters_t *p
     if( status != PSA_SUCCESS )
         goto exit;
 
-    status = psa_hash_finish( &op, out,
-                              MBEDTLS_LMOTS_N_HASH_LEN(params->type),
+    status = psa_hash_finish( &op, temp,
+                              MBEDTLS_LMOTS_N_HASH_LEN_MAX,
                               &output_hash_len );
     if( status != PSA_SUCCESS )
         goto exit;
 
+    memcpy( out, temp, MBEDTLS_LMOTS_N_HASH_LEN( params->type ) );
```

Comment:
As an API designer, I'm now wondering whether to add SHA-256/192 and SHAKE256/192 as distinct algorithms in the PSA API.

On the one hand, they can be used as hash algorithms, and the caller requesting a hash shouldn't really have to know how they're calculated. There's a precedent with MAC and AEAD where truncated MACs/tags have distinct algorithm encodings. There are multiple places that need to calculate the truncated hash — several already in `lmots.c` and `lms.c`, and there would be more if we implemented HSS and XMSS.

On the other hand, adding truncated values makes the API messier, and the encoding of algorithms makes is inconvenient to have more than 128 values total (8 bits, with one preferably reserved for vendor-specific encodings). And technically [SP 800-208](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-208.pdf) doesn't define them as hash algorithms, but as truncations of hash algorithms.

So I'm currently leaning towards no, don't define new encodings. But open to changing my mind.

Reply:
Good points on both sides of the argument. Personally I'd be in favor of adding them to PSA API.